### ISSUE TITLE
python dependencies are affecting the CI of multiple CKF repositories

### ISSUE THREAD

**Issue by DnPlas** [canonical/bundle-kubeflow#648]
In the past few weeks, a number of charm and tests dependencies (python packages) have been affecting the health of most of the CIs of the charms that belong to the CKF bundle. We have identified two main issues:

## `pyyaml` 6.0 fails to install and returns `AttributeError: cython_sources`

#### Description
`pyyaml` is a dependency of `python-libjuju`, which before v2.9.44 pinned `pyyaml>=5.1.2,<=6.0`, which did not allow  `pyyaml 6.0.1` to be installed, preventing a fix for https://github.com/yaml/pyyaml/issues/601 to be pulled in.
For more details, please go [here](https://github.com/juju/python-libjuju/issues/913).

The issue can be identified when installing pyyaml, pip will throw the following message:

<details>
  <summary>Pip install output</summary>

```

2023-07-18 10:03:06.452 ::    :: Collecting pyyaml==6.0
2023-07-18 10:03:06.501 ::    ::   Downloading PyYAML-6.0.tar.gz (124 kB)
2023-07-18 10:03:06.948 ::    ::   Installing build dependencies: started
2023-07-18 10:07:10.666 ::    ::   Installing build dependencies: still running...
2023-07-18 10:11:35.589 ::    ::   Installing build dependencies: still running...
2023-07-18 10:13:17.459 ::    ::   Installing build dependencies: still running...
2023-07-18 10:14:35.583 ::    ::   Installing build dependencies: still running...
2023-07-18 10:16:05.758 ::    ::   Installing build dependencies: still running...
2023-07-18 10:17:50.556 ::    ::   Installing build dependencies: still running...
2023-07-18 10:18:50.586 ::    ::   Installing build dependencies: still running...
2023-07-18 10:19:52.667 ::    ::   Installing build dependencies: still running...
2023-07-18 10:20:04.634 ::    ::   Installing build dependencies: finished with status 'done'
2023-07-18 10:20:04.706 ::    ::   Getting requirements to build wheel: started
2023-07-18 10:20:05.690 ::    ::   Getting requirements to build wheel: finished with status 'error'
2023-07-18 10:20:05.695 ::    ::   ERROR: Command errored out with exit status 1:
2023-07-18 10:20:05.697 ::    ::    command: /home/runner/work/kfp-operators/kfp-operators/charms/kfp-viewer/build/parts/charm/build/staging-venv/bin/python3 /tmp/tmpojz_j9um get_requires_for_build_wheel /tmp/tmpg5tg6lyl
2023-07-18 10:20:05.697 ::    ::        cwd: /tmp/pip-install-eanq3h0w/pyyaml
2023-07-18 10:20:05.698 ::    ::   Complete output (48 lines):
2023-07-18 10:20:05.698 ::    ::   running egg_info
2023-07-18 10:20:05.698 ::    ::   writing lib/PyYAML.egg-info/PKG-INFO
2023-07-18 10:20:05.698 ::    ::   writing dependency_links to lib/PyYAML.egg-info/dependency_links.txt
2023-07-18 10:20:05.698 ::    ::   writing top-level names to lib/PyYAML.egg-info/top_level.txt
2023-07-18 10:20:05.698 ::    ::   Traceback (most recent call last):
2023-07-18 10:20:05.698 ::    ::     File "/tmp/tmpojz_j9um", line 280, in <module>
2023-07-18 10:20:05.698 ::    ::       main()
2023-07-18 10:20:05.698 ::    ::     File "/tmp/tmpojz_j9um", line 263, in main
2023-07-18 10:20:05.698 ::    ::       json_out['return_val'] = hook(**hook_input['kwargs'])
2023-07-18 10:20:05.698 ::    ::     File "/tmp/tmpojz_j9um", line 114, in get_requires_for_build_wheel
2023-07-18 10:20:05.698 ::    ::       return hook(config_settings)
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 341, in get_requires_for_build_wheel
2023-07-18 10:20:05.699 ::    ::       return self._get_build_requires(config_settings, requirements=['wheel'])
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 323, in _get_build_requires
2023-07-18 10:20:05.699 ::    ::       self.run_setup()
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 338, in run_setup
2023-07-18 10:20:05.699 ::    ::       exec(code, locals())
2023-07-18 10:20:05.699 ::    ::     File "<string>", line 288, in <module>
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/__init__.py", line 107, in setup
2023-07-18 10:20:05.699 ::    ::       return distutils.core.setup(**attrs)
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 185, in setup
2023-07-18 10:20:05.699 ::    ::       return run_commands(dist)
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
2023-07-18 10:20:05.699 ::    ::       dist.run_commands()
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 969, in run_commands
2023-07-18 10:20:05.699 ::    ::       self.run_command(cmd)
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/dist.py", line 1234, in run_command
2023-07-18 10:20:05.699 ::    ::       super().run_command(command)
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
2023-07-18 10:20:05.699 ::    ::       cmd_obj.run()
2023-07-18 10:20:05.699 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/command/egg_info.py", line 314, in run
2023-07-18 10:20:05.699 ::    ::       self.find_sources()
2023-07-18 10:20:05.700 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/command/egg_info.py", line 322, in find_sources
2023-07-18 10:20:05.700 ::    ::       mm.run()
2023-07-18 10:20:05.700 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/command/egg_info.py", line 551, in run
2023-07-18 10:20:05.700 ::    ::       self.add_defaults()
2023-07-18 10:20:05.700 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/command/egg_info.py", line 589, in add_defaults
2023-07-18 10:20:05.700 ::    ::       sdist.add_defaults(self)
2023-07-18 10:20:05.700 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/command/sdist.py", line 104, in add_defaults
2023-07-18 10:20:05.700 ::    ::       super().add_defaults()
2023-07-18 10:20:05.700 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/_distutils/command/sdist.py", line 251, in add_defaults
2023-07-18 10:20:05.700 ::    ::       self._add_defaults_ext()
2023-07-18 10:20:05.700 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/_distutils/command/sdist.py", line 336, in _add_defaults_ext
2023-07-18 10:20:05.700 ::    ::       self.filelist.extend(build_ext.get_source_files())
2023-07-18 10:20:05.700 ::    ::     File "<string>", line 204, in get_source_files
2023-07-18 10:20:05.700 ::    ::     File "/tmp/pip-build-env-izjg03kl/overlay/lib/python3.8/site-packages/setuptools/_distutils/cmd.py", line 107, in __getattr__
2023-07-18 10:20:05.700 ::    ::       raise AttributeError(attr)
2023-07-18 10:20:05.700 ::    ::   AttributeError: cython_sources
2023-07-18 10:20:05.700 ::    ::   ----------------------------------------

```
</details>


#### Affects
Any type of code that uses `python-libjuju=<2.9.43`, specially of course, integration tests.

#### Solution
Bump the version of `python-libjuju>=2.9.44`, making sure that `pyyaml>6.0` is installed.

## Fail to build charm `RuntimeError: Failed to build charm`

#### Description
Although this is a generic message when a `charmcraft pack` command fails, we are focusing on the following error message:

<details>
  <summary>charmcraft pack logs</summary>

```
::    :: Collecting rpds-py==0.9.2 (from -r requirements.txt (line 75))                                                                                                                                                                 
::    ::   Using cached rpds_py-0.9.2.tar.gz (16 kB)                                                                                                                                                                                    
::    ::   Installing build dependencies: started                                                                                                                                                                                       
::    ::   Installing build dependencies: finished with status 'error'                                                                                                                                                                  
::    ::   error: subprocess-exited-with-error                                                                                                                                                                                          
::    ::                                                                                                                                                                                                                                
::    ::   × pip subprocess to install build dependencies did not run successfully.                                                                                                                                                     
::    ::   │ exit code: 1                                                                                                                                                                                                               
::    ::   ╰─> [48 lines of output]                                                                                                                                                                                                     
::    ::       Collecting maturin<2.0,>=1.0                                                                                                                                                                                             
::    ::         Using cached maturin-1.1.0.tar.gz (181 kB)                                                                                                                                                                             
::    ::         Installing build dependencies: started                                                                                                                                                                                 
::    ::         Installing build dependencies: finished with status 'done'                                                                                                                                                             
::    ::         Getting requirements to build wheel: started                                                                                                                                                                           
::    ::         Getting requirements to build wheel: finished with status 'done'                                                                                                                                                       
::    ::         Preparing metadata (pyproject.toml): started                                                                                                                                                                           
::    ::         Preparing metadata (pyproject.toml): finished with status 'done'                                                                                                                                                       
::    ::       Collecting tomli>=1.1.0 (from maturin<2.0,>=1.0)                                                                                                                                                                         
::    ::         Using cached tomli-2.0.1-py3-none-any.whl                                                                                                                                                                              
::    ::       Building wheels for collected packages: maturin                                                                                                                                                                          
::    ::         Building wheel for maturin (pyproject.toml): started                                                                                                                                                                   
::    ::         Building wheel for maturin (pyproject.toml): finished with status 'error'                                                                                                                                              
::    ::         error: subprocess-exited-with-error                                                                                                                                                                                    
::    ::                                                                                                                                                                                                                                
::    ::         × Building wheel for maturin (pyproject.toml) did not run successfully.                                                                                                                                                
::    ::         │ exit code: 1                                                                                                                                                                                                         
::    ::         ╰─> [24 lines of output]                                                                                                                                                                                               
::    ::             running bdist_wheel                                                                                                                                                                                                
::    ::             running build                                                                                                                                                                                                      
::    ::             running build_py                                                                                                                                                                                                   
::    ::             creating build                                                                                                                                                                                                     
::    ::             creating build/lib.linux-x86_64-cpython-38                                                                                                                                                                         
::    ::             creating build/lib.linux-x86_64-cpython-38/maturin                                                                                                                                                                 
::    ::             copying maturin/__main__.py -> build/lib.linux-x86_64-cpython-38/maturin                                                                                                                                           
::    ::             copying maturin/__init__.py -> build/lib.linux-x86_64-cpython-38/maturin                                                                                                                                           
::    ::             copying maturin/import_hook.py -> build/lib.linux-x86_64-cpython-38/maturin                                                                                                                                        
::    ::             running egg_info                                                                                                                                                                                                   
::    ::             creating maturin.egg-info                                                                                                                                                                                          
::    ::             writing maturin.egg-info/PKG-INFO                                                                                                                                                                                  
::    ::             writing dependency_links to maturin.egg-info/dependency_links.txt                                                                                                                                                  
::    ::             writing requirements to maturin.egg-info/requires.txt                                                                                                                                                              
::    ::             writing top-level names to maturin.egg-info/top_level.txt                                                                                                                                                          
::    ::             writing manifest file 'maturin.egg-info/SOURCES.txt'                                                                                                                                                               
::    ::             reading manifest file 'maturin.egg-info/SOURCES.txt'                                                                                                                                                               
::    ::             reading manifest template 'MANIFEST.in'                                                                                                                                                                            
::    ::             warning: no files found matching '*.json' under directory 'src/python_interpreter'                                                                                                                                 
::    ::             writing manifest file 'maturin.egg-info/SOURCES.txt'                                                                                                                                                               
::    ::             running build_ext                                                                                                                                                                                                  
::    ::             running build_rust                                                                                                                                                                                                 
::    ::             cargo build --manifest-path Cargo.toml --message-format=json-render-diagnostics --release -v --no-default-features --locked                                                                                        
::    ::             error: Unable to execute 'cargo' - this package requires Rust to be installed and cargo to be on the PATH                                                                                                          
::    ::             [end of output]                                                                                                                                                                                                    
::    ::                                                                                                                                                                                                                                
::    ::         note: This error originates from a subprocess, and is likely not a problem with pip.                                                                                                                                   
::    ::         ERROR: Failed building wheel for maturin                                                                                                                                                                               
::    ::       Failed to build maturin                                                                                                                                                                                                  
::    ::       ERROR: Could not build wheels for maturin, which is required to install pyproject.toml-based projects                                                                                                                    
::    ::       [end of output]                                                                                                                                                                                                          
::    ::                                                                                                                                                                                                                                
::    ::   note: This error originates from a subprocess, and is likely not a problem with pip.                                                                                                                                         
::    :: error: subprocess-exited-with-error                                                                                                                                                                                            
::    ::                                                                                                                                                                                                                                
::    :: × pip subprocess to install build dependencies did not run successfully.                                                                                                                                                       
::    :: │ exit code: 1                                                                                                                                                                                                                 
::    :: ╰─> See above for output.                                                                                                                                                                                                      
::    ::                                                                                                                                                                                                                                
::    :: note: This error originates from a subprocess, and is likely not a problem with pip.                                                                                                                                           
:: Traceback (most recent call last):                                                                                                                                                                                                   
::   File "/snap/charmcraft/1349/lib/charmcraft/charm_builder.py", line 435, in <module>                                                                                                                                                
::     main()                                                                                                                                                                                                                           
::   File "/snap/charmcraft/1349/lib/charmcraft/charm_builder.py", line 430, in main                                                                                                                                                    
::     builder.build_charm()                                                                                                                                                                                                            
::   File "/snap/charmcraft/1349/lib/charmcraft/charm_builder.py", line 103, in build_charm                                                                                                                                             
::     self.handle_dependencies()                                                                                                                                                                                                       
::   File "/snap/charmcraft/1349/lib/charmcraft/charm_builder.py", line 317, in handle_dependencies                                                                                                                                     
::     self._install_dependencies(staging_venv_dir)                                                                                                                                                                                     
::   File "/snap/charmcraft/1349/lib/charmcraft/instrum.py", line 152, in _f                                                                                                                                                            
::     return func(*args, **kwargs)                                                                                                                                                                                                     
::   File "/snap/charmcraft/1349/lib/charmcraft/charm_builder.py", line 271, in _install_dependencies                                                                                                                                   
::     _process_run(cmd)                                                                                                                                                                                                                
::   File "/snap/charmcraft/1349/lib/charmcraft/charm_builder.py", line 371, in _process_run                                                                                                                                            
::     raise RuntimeError(f"Subprocess command {cmd} execution failed with retcode {retcode}")                                                                                                                                          
:: RuntimeError: Subprocess command ['/root/parts/charm/build/staging-venv/bin/pip3', 'install', '--upgrade', '--no-binary', ':all:', '--requirement=requirements.txt'] execution failed with retcode 1                                 
Parts processing error: Failed to run the build script for part 'charm'.                                                                                                                                                                
Failed to build charm for bases index '0'.                                                                                                                                                                                              
Full execution log: '/home/ubuntu/.local/state/charmcraft/log/charmcraft-20230724-144949.605632.log'     
```

</details>

This error is caused by a `jsonschema` v4.18.x dependency that requires `rustc` and `cargo` to be present in the charm to be built. Please refer to https://github.com/python-jsonschema/jsonschema/issues/1117 and https://github.com/python-jsonschema/jsonschema/issues/1114 for more details on the issue.

#### Affects

This error will be present **ONLY when bumping `jsonschema` to 4.18.x** in a charm's `requirements.txt` file.

> NOTE: this error is thrown by pip when attempting to install `jsonschema` 4.18.x on any environment that does not have `rustc` and `cargo`. Beware that this is not a charm specific thing, but an environment thing and can happen anywhere.

#### Solution
For the majority of the cases, `jsonschema` is introduced as a dependency of `SDI`, therefore using the latest version (0.7.0) of that package and updating the `requirements.txt` file should solve the `charmcraft pack` issue.

> NOTE: for charms that are not compatible with the latest version of `SDI`, pinning both `SDI` and the `jsonschema` in `requirements.txt` is the way to fix this.

In the case this is happens outside of a charm that uses SDI, pinning `jsonschema` may fix the issue.

**Comment by i-chvets** [comment]
While reseaching the solution for Dex Auth requirements update ([PR](https://github.com/canonical/dex-auth-operator/pull/141) it was confirmed that the above two solutions proposed in this issue are not applicable.
The problem lies in the fact that SDI 0.3.5 used in Dex requires PyYAML 5.4. This cannot be addressed with solutions proposed above.
One of the possible approaches is to relax PyYAML requirements in SDI itself and relased SDI 0.3.6 with PyYAML 6.0.1.
This will allow to fix CI in Dex and other charms that require SDI of at least 0.3.5

**Comment by DnPlas** [comment]
@i-chvets we have a new version of SDI that should be compatible with `dex-auth` from `track/2.31`. Do you mind trying installing SDI 0.3.6?

**Comment by i-chvets** [comment]
> @i-chvets we have a new version of SDI that should be compatible with `dex-auth` from `track/2.31`. Do you mind trying installing SDI 0.3.6?

It worked like a charm. Fix is ready for review https://github.com/canonical/dex-auth-operator/pull/143

**Comment by DnPlas** [comment]
The proposed solution in the description of the issue has been successfully tested in affected repos in `main`. All of them should now have a PR either merged or in process of being merged to fix this. I'm not listing them manually here, but I have made sure I've added the note: `Part of #648` so that everyone knows the commit to be merged is related to this issue.

This issue may also affect version branches (`track/<version>`) that the team still supports, so I'm not closing the issue yet to have some reference for those branches as well.
